src/box/lua/console.lua

```diff
@@ -436,6 +524,18 @@ local function local_eval(storage, line)
     if not res[1] and not in_txn_before and ffi.C.box_txn() then
         box.rollback()
     end
+    -- Check trace is correct. Skip check for non ClienError.
+    if not res[1] and check_trace then
+        local err = res[2]
+        if err.type == nil or err.trace == nil then
+            return format(false, string.format('Non box.error: %s', err))
+        end
+        local trace = err.trace[1]
+        if not table.equals(trace, proper_trace) then
+            return format(false, string.format('Unexpected trace: %s:%d',
+                                               trace.file, trace.line))
```

Comment:
I think it's better to append the invalid trace to the original error object instead of overriding it.

Reply:
Fixed.

```diff
diff --git a/src/box/lua/console.lua b/src/box/lua/console.lua
index 7a906036f..0f3a75a1c 100644
--- a/src/box/lua/console.lua
+++ b/src/box/lua/console.lua
@@ -532,8 +532,9 @@ local function local_eval(storage, line)
         end
         local trace = err.trace[1]
         if not table.equals(trace, proper_trace) then
-            return format(false, string.format('Unexpected trace: %s:%d',
-                                               trace.file, trace.line))
+            return format(false,
+                          string.format('Unexpected trace %s:%d for error: %s',
+                                        trace.file, trace.line, err))
        end
     end
     -- specify length to preserve trailing nils
```

Reply:
Let's just put it in a table?

```lua
return format(false, {err, trace})
```

Reply:
Done. Also make check run only for test build.

```diff
--- a/src/box/lua/console.lua
+++ b/src/box/lua/console.lua
@@ -501,17 +501,15 @@ local function local_eval(storage, line)
     -- It is hard to fix box.error trace for all API at once so we will fix it
     -- per module basis.
     --
-    if not res[1] and
+    if tarantool.build.test_build and not res[1] and
        tarantool._internal.trace_check_is_required(res[2].info.short_src) then
         local err = res[2].err
         if not box.error.is(err) then
-            return format(false, string.format('Non box.error: %s', err))
+            return format(false, {err, 'Warning, box error expected'})
         end
         local trace = err.trace[1]
         if not table.equals(trace, wrapped_trace) then
-            return format(false,
-                          string.format('Unexpected trace %s:%d for error: %s',
-                                        trace.file, trace.line, err))
+            return format(false, {err, 'Warning, unexpected trace', trace})
        end
     end
     if not res[1] then
```